src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -413,34 +414,46 @@ public bool TryCreateType(ITypeSymbol symbol, [NotNullWhen(true)] out CSharpType
 
         public bool TryCreateType(ITypeSymbol symbol, Func<System.Type, bool> validator, [NotNullWhen(true)] out CSharpType? type)
         {
-            type = null;
-            INamedTypeSymbol? namedTypeSymbol = symbol as INamedTypeSymbol;
-            if (namedTypeSymbol == null)
+            if (TryCreateTypeForIArrayTypeSymbol(symbol, out type))
+            {
+                return true;
+            }
+
+            // We can only handle IArrayTypeSymbol of framework type and INamedTypeSymbol for now since CSharpType can't represent other types such as IArrayTypeSymbol of user types
+            // Instead of throwing an exception, wihch causes more side effects, we just return false and let the caller handle it.
+            if (symbol is not INamedTypeSymbol namedTypeSymbol)
             {
-                throw new InvalidCastException($"Unexpected type {symbol}");
+                return false;
             }
 
+            return TryCreateTypeForINamedTypeSymbol(symbol, validator, out type, namedTypeSymbol);
```

Comment:
could we change this to a switch so that it would be more clear and much easier for us to add more case
such as:
```csharp
=> symbol switch
{
IArrayTypeSymbol arraySymbol => TryCreateTypeForIArrayTypeSymbol(arraySymbol, out type),
INamedTypeSymbol namedTypeSymbol => TryCreateTypeForINamedTypeSymbol(namedTypeSymbol, validator, out type),
_ => false,
}
```

Reply:
And we could change the signature of `TryCreateTypeForIArrayTypeSymbol` to only accept `IArraySymbol`

Reply:
sounds good